File: go/othello/board.go
//==========================================================
// Board
// Defines the game board
// Akseli Lukkarila
// 2019-2026
//==========================================================

package othello

import (
	"fmt"
	"sort"
	"strconv"
	"strings"

	mapset "github.com/deckarep/golang-set/v2"
	"github.com/logrusorgru/aurora/v4"
)

const (
	up    = 1
	down  = -1
	left  = -1
	right = 1
	still = 0
)

// All possible step directions for a square on the board.
var stepDirections = [8]Step{
	{X: down, Y: left},
	{X: down, Y: right},
	{X: down, Y: still},
	{X: still, Y: left},
	{X: still, Y: right},
	{X: up, Y: left},
	{X: up, Y: right},
	{X: up, Y: still},
}

// Board Handles game board state and logic.
type Board struct {
	board        []Disk
	emptySquares mapset.Set[Square]
	indices      []int
	size         int
}

// NewBoard Initialize a new game board with the given size.
func NewBoard(size int) Board {
	board := initBoard(size)

	// Index list (0...size) to avoid repeating same range in loops.
	indices := make([]int, size)
	for i := range indices {
		indices[i] = i
	}

	// Keep track of empty squares on board to avoid checking already filled positions.
	emptySquares := initEmptySquares(size, board)

	return Board{
		board:        board,
		emptySquares: emptySquares,
		indices:      indices,
		size:         size,
	}
}

// CanPlay Return true if board contains empty squares.
func (b *Board) CanPlay() bool {
	return b.emptySquares.Cardinality() > 0
}

// PlaceDisk Update board for given disk placement.
func (b *Board) PlaceDisk(chosenMove *Move) {
	start := chosenMove.Square
	disk, ok := b.getSquare(&start)
	if !ok {
		panic(fmt.Sprintf("Invalid coordinates: %s", start))
	}
	if disk != Empty {
		panic(fmt.Sprintf("Trying to place disk to an occupied square: %s!", start))
	}
	b.setSquare(&start, chosenMove.Disk)
	b.emptySquares.Remove(start)
	for _, square := range chosenMove.AffectedSquares() {
		b.setSquare(&square, chosenMove.Disk)
	}
}

// PossibleMoves Returns a list of possible moves for the given player.
func (b *Board) PossibleMoves(disk Disk) []Move {
	var moves []Move
	opposingDisk := disk.Opponent()
	for square := range b.emptySquares.Iter() {
		var value int
		var directions []Direction
		for _, step := range stepDirections {
			pos := square.Add(step)
			// Next square in this direction needs to be the opposing disk
			squareDisk, _ := b.getSquare(&pos)
			if squareDisk != opposingDisk {
				continue
			}
			var numSteps int
			// Keep stepping over opponents disks
			for squareDisk == opposingDisk {
				numSteps++
				pos = pos.Add(step)
				squareDisk, _ = b.getSquare(&pos)
			}
			// Valid move only if a line of opposing disks ends with own disk
			if squareDisk == disk {
				directions = append(directions, Direction{step, numSteps})
				value += numSteps
			}
		}
		if value > 0 {
			moves = append(moves, Move{
				Square:     square,
				Disk:       disk,
				Value:      value,
				Directions: directions,
			})
		}
	}
	if len(moves) > 0 {
		sort.Sort(MovesDescending(moves))
	}
	return moves
}

// PrintPossibleMoves Print board with available move coordinates and the resulting points gained.
func (b *Board) PrintPossibleMoves(moves []Move) {
	fmt.Println(aurora.Yellow(fmt.Sprintf("  Possible moves (%d):", len(moves))))
	// Convert board from Disk enums to strings
	formattedBoard := make([]string, len(b.board))
	for i, disk := range b.board {
		formattedBoard[i] = disk.BoardCharWithColor()
	}
	// Add possible moves to board
	for _, possibleMove := range moves {
		index := b.squareIndex(&possibleMove.Square)
		formattedBoard[index] = aurora.Yellow(strconv.Itoa(possibleMove.Value)).String()
		fmt.Printf("  %s\n", possibleMove)
	}
	// Print board with move positions
	fmt.Print("    ")
	for _, i := range b.indices {
		fmt.Printf(" %s", aurora.Bold(strconv.Itoa(i)))
	}
	for _, y := range b.indices {
		fmt.Printf("\n  %s", aurora.Bold(strconv.Itoa(y)))
		for _, x := range b.indices {
			fmt.Printf(" %s", formattedBoard[y*b.size+x])
		}
	}
	fmt.Println()
}

// PrintScore Print current score for both players.
func (b *Board) PrintScore() {
	black, white := b.playerScores()
	fmt.Printf("\n%s\n", b)
	fmt.Printf("Score: %d | %d\n", aurora.Magenta(black), aurora.Cyan(white))
}

// Result Returns the winning disk colour. Empty indicates a draw.
func (b *Board) Result() Disk {
	totalScore := b.score()
	switch {
	case totalScore > 0:
		return White
	case totalScore < 0:
		return Black
	default:
		return Empty
	}
}

// LogEntry Get board status string for game log.
func (b *Board) LogEntry() string {
	result := ""
	for _, disk := range b.board {
		result += disk.BoardChar()
	}
	return result
}

// Check that the given coordinates are valid (inside the board).
func (b *Board) checkCoordinates(x, y int) bool {
	return x >= 0 && x < b.size && y >= 0 && y < b.size
}

// Check that the given square is valid (inside the board).
func (b *Board) checkSquare(square *Square) bool {
	return b.checkCoordinates(square.X, square.Y)
}

// Returns the state of the board (empty, white, black) at the given square.
// The boolean return value will be false if the square is outside the board.
func (b *Board) getSquare(square *Square) (Disk, bool) {
	if b.checkSquare(square) {
		index := b.squareIndex(square)
		return b.board[index], true
	}
	return Empty, false
}

// Map square to board index.
func (b *Board) squareIndex(square *Square) int {
	return square.Y*b.size + square.X
}

// Count and return the number of black and white disks.
func (b *Board) playerScores() (int, int) {
	var black, white int
	for _, disk := range b.board {
		switch disk {
		case Black:
			black++
		case White:
			white++
		}
	}
	return black, white
}

// Returns the total score.
// Positive value means more white disks and negative means more black disks.
func (b *Board) score() int {
	var sum int
	for _, disk := range b.board {
		sum += int(disk)
	}
	return sum
}

// Sets the given square to the given value.
func (b *Board) setSquare(square *Square, disk Disk) {
	if !b.checkSquare(square) {
		panic(fmt.Sprintf("Invalid coordinates: %s", square))
	}
	index := b.squareIndex(square)
	b.board[index] = disk
}

// Initialize game board with starting disk positions.
func initBoard(size int) []Disk {
	// Initialize game board with empty disks
	board := make([]Disk, size*size)
	for i := range board {
		board[i] = Empty
	}
	// Set starting positions
	var row int
	if size%2 == 0 {
		row = (size - 1) / 2
	} else {
		row = (size-1)/2 - 1
	}
	col := size / 2
	board[row*size+row] = White
	board[row*size+col] = Black
	board[col*size+row] = Black
	board[col*size+col] = White
	return board
}

// Initialize empty squares for the board.
func initEmptySquares(size int, board []Disk) mapset.Set[Square] {
	emptySquares := mapset.NewSet[Square]()
	for i, disk := range board {
		if disk == Empty {
			emptySquares.Add(Square{X: i % size, Y: i / size})
		}
	}
	return emptySquares
}

// String Format game board to string.
func (b *Board) String() string {
	// Horizontal indices
	numbers := make([]string, len(b.indices))
	for i, index := range b.indices {
		numbers[i] = strconv.Itoa(index)
	}
	text := fmt.Sprintf("  %s", aurora.Bold(strings.Join(numbers, " ")))
	for _, y := range b.indices {
		// Vertical index
		text += fmt.Sprintf("\n%s", aurora.Bold(strconv.Itoa(y)))
		// Row values
		for _, x := range b.indices {
			disk := b.board[y*b.size+x]
			text += fmt.Sprintf(" %s", disk.BoardCharWithColor())
		}
	}
	return text
}

File: csharp/Othello/Board.cs
//==========================================================
// Class Board
// Defines the game board
// Akseli Lukkarila
// 2019-2026
//==========================================================

using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;

namespace Othello;

/// <summary>Handles game board state and logic.</summary>
internal sealed class Board {
    private const int Up = 1;
    private const int Down = -1;
    private const int Left = -1;
    private const int Right = 1;
    private const int Still = 0;

    /// <summary>All possible step directions for a square on the board.</summary>
    private static readonly Step[] StepDirections =
    [
        new(Down, Left),
        new(Down, Right),
        new(Down, Still),
        new(Still, Left),
        new(Still, Right),
        new(Up, Left),
        new(Up, Right),
        new(Up, Still),
    ];

    private readonly List<Disk> _board;
    private readonly List<Square> _emptySquares;
    private readonly List<int> _indices;
    internal readonly int Size;

    /// <summary>Initialize a new board for the given board size.</summary>
    public Board(int size) {
        _board = InitBoard(size);

        // Index list (0...size) to avoid repeating same range in loops.
        _indices = [.. Enumerable.Range(0, size)];

        // Keep track of empty squares on board to avoid checking already filled positions.
        _emptySquares = InitEmptySquares(size, _board);

        Size = size;
    }

    /// <summary>Return true if board contains empty squares.</summary>
    public bool CanPlay() {
        return _emptySquares.Count != 0;
    }

    /// <summary>Update board for given disk placement.</summary>
    public void PlaceDisk(Move chosenMove) {
        Square start = chosenMove.Square;
        Disk? oldDisk =
            GetSquare(start)
            ?? throw new ArgumentException($"Invalid coordinates: {start}");
        if (oldDisk != Disk.Empty) {
            throw new ArgumentException($"Trying to place disk to an occupied square: {start}!");
        }
        SetSquare(start, chosenMove.Disk);
        _emptySquares.Remove(start);
        foreach (Square square in chosenMove.AffectedSquares()) {
            SetSquare(square, chosenMove.Disk);
        }
    }

    /// <summary>Returns a list of possible moves for the given player.</summary>
    public List<Move> PossibleMoves(Disk disk) {
        List<Move> moves = [];
        Disk opposingDisk = disk.Opponent();
        foreach (Square square in _emptySquares) {
            int value = 0;
            List<Direction> directions = [];
            foreach (Step step in StepDirections) {
                Square pos = square + step;
                // Next square in this direction needs to be the opposing disk
                if (GetSquare(pos) != opposingDisk) {
                    continue;
                }
                int numSteps = 0;
                // Keep stepping over opponents disks
                while (GetSquare(pos) == opposingDisk) {
                    ++numSteps;
                    pos += step;
                }
                // Valid move only if a line of opposing disks ends with own disk
                if (GetSquare(pos) == disk) {
                    directions.Add(new Direction(step, numSteps));
                    value += numSteps;
                }
            }
            if (value > 0) {
                moves.Add(new Move(square, disk, value, directions));
            }
        }
        moves.Sort();
        return moves;
    }

    /// <summary>Print board with available move coordinates and the resulting points gained.</summary>
    public void PrintPossibleMoves(IReadOnlyCollection<Move> moves) {
        ColorPrint.PrintYellow($"  Possible moves ({moves.Count}):");
        // Convert board from Disk enums to strings
        List<string> formattedBoard = [.. _board.Select(disk => disk.BoardCharWithColor())];

        // Add possible moves to board
        foreach (Move possibleMove in moves) {
            int index = SquareIndex(possibleMove.Square);
            formattedBoard[index] = ColorPrint.Get(possibleMove.Value, Color.Yellow);
            Console.WriteLine($"  {possibleMove}");
        }
        // Print board with move positions
        Console.Write("    ");
        foreach (int i in _indices) {
            Console.Write($" {ColorPrint.Bold(i)}");
        }
        foreach (int y in _indices) {
            Console.Write($"\n  {ColorPrint.Bold(y)}");
            foreach (int x in _indices) {
                Console.Write($" {formattedBoard[y * Size + x]}");
            }
        }
        Console.WriteLine();
    }

    /// <summary>Print current score for both players.</summary>
    public void PrintScore() {
        (int black, int white) = PlayerScores();
        Console.WriteLine($"\n{this}");
        Console.WriteLine(
            $"Score: {ColorPrint.Get(black, Disk.Black.DiskColor())} | "
                + $"{ColorPrint.Get(white, Disk.White.DiskColor())}"
        );
    }

    /// <summary>Returns the winning disk colour. Empty indicates a draw.</summary>
    public Disk Result() {
        int totalScore = Score();
        if (totalScore == 0) {
            return Disk.Empty;
        }
        return totalScore > 0 ? Disk.White : Disk.Black;
    }

    /// <summary>Get board status string for game log.</summary>
    public string LogEntry() {
        return _board.Aggregate(
            new StringBuilder(),
            (accumulator, disk) => accumulator.Append(disk.BoardChar()),
            accumulator => accumulator.ToString()
        );
    }

    /// <summary>Check that the given coordinates are valid (inside the board).</summary>
    private bool CheckCoordinates(int x, int y) {
        return 0 <= x && x < Size && 0 <= y && y < Size;
    }

    /// <summary>Check that the given square is valid (inside the board).</summary>
    private bool CheckSquare(Square square) {
        return CheckCoordinates(square.X, square.Y);
    }

    /// <summary>Returns the state of the board (empty, white, black) at the given square.</summary>
    internal Disk? GetSquare(Square square) {
        if (CheckSquare(square)) {
            return _board[SquareIndex(square)];
        }
        return null;
    }

    /// <summary>Map square to board index.</summary>
    private int SquareIndex(Square square) {
        return square.Y * Size + square.X;
    }

    /// <summary>Count and return the number of black and white disks.</summary>
    internal (int black, int white) PlayerScores() {
        int black = 0;
        int white = 0;
        foreach (Disk disk in _board) {
            switch (disk) {
                case Disk.Black:
                    ++black;
                    break;
                case Disk.White:
                    ++white;
                    break;
                case Disk.Empty:
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }
        return (black, white);
    }

    /// <summary>
    /// Returns the total score.
    /// Positive value means more white disks and negative means more black disks.
    /// </summary>
    internal int Score() {
        return _board.Sum(static disk => (int)disk);
    }

    /// <summary>Sets the given square to the given value.</summary>
    private void SetSquare(Square square, Disk disk) {
        if (!CheckSquare(square)) {
            throw new ArgumentException($"Invalid coordinates: {square}");
        }
        _board[SquareIndex(square)] = disk;
    }

    /// <summary>Initialize game board with starting disk positions.</summary>
    private static List<Disk> InitBoard(int size) {
        // Initialize game board with empty disks
        List<Disk> board = [.. Enumerable.Repeat(Disk.Empty, size * size)];
        // Set starting positions
        int row = size % 2 == 0 ? (size - 1) / 2 : (size - 1) / 2 - 1;
        int col = size / 2;
        board[row * size + row] = Disk.White;
        board[row * size + col] = Disk.Black;
        board[col * size + row] = Disk.Black;
        board[col * size + col] = Disk.White;
        return board;
    }

    /// <summary>Initialize empty squares for the board.</summary>
    private static List<Square> InitEmptySquares(int size, List<Disk> board) {
        return
        [
            .. Enumerable
                .Range(0, board.Count)
                .Where(i => board[i] == Disk.Empty)
                .Select(i => new Square(i % size, i / size)),
        ];
    }

    /// <summary>Format game board to string</summary>
    public override string ToString() {
        // Horizontal indices
        string text = $"  {ColorPrint.Bold(string.Join(" ", _indices))}";
        foreach (int y in _indices) {
            // Vertical index
            text += $"\n{ColorPrint.Bold(y)}";
            // Row values
            foreach (int x in _indices) {
                text += $" {_board[y * Size + x].BoardCharWithColor()}";
            }
        }
        return text;
    }
}
